### PR TITLE
[server] terminate ws on shutdown

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Terminate websocket connections on SIGINT.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 9d5e1cc</samp>

This pull request enhances the cleanup and shutdown logic for the server and the websocket handler, using the `Disposable` pattern and adding more logging. It also fixes some minor bugs and import issues in `server.ts` and `ws-handler.ts`.

</details>

## Related Issue(s)

Fixes EXP-476

## How to test

- login into the preview environment
- start a workspace on this PR and set the kubectx to the preview
- follow the logs of the server pod
- delete the server pod and notice that it immediately terminates (you can tell from the logs)

kubectll will still show the old pod in terminating while the new pod is initializing.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
        <li><b>🏷️ Name</b> - se-cancel-ws</li>
        <li><b>🔗 URL</b> - <a href="https://se-cancel-ws.preview.gitpod-dev.com/workspaces" target="_blank">se-cancel-ws.preview.gitpod-dev.com/workspaces</a>.</li>
        <li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
        <li><b>📦 Version</b> - </li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by geropl** [comment]
This does overlap with https://github.com/gitpod-io/gitpod/pull/18548 , right? :thinking:

**Comment by svenefftinge** [comment]
duplicate of #18548

File: components/server/src/express/ws-handler.ts
Comment (geropl) [inline components/server/src/express/ws-handler.ts:71]: We do similar bookkeeping here: https://github.com/gitpod-io/gitpod/blob/478a75e744a642d9b764de37cfae655bc8b29dd5/components/server/src/express/ws-connection-handler.ts#L21.
We could register the disposables there as well.

**Comment by geropl** [comment]
triggered a re-deploy

File: components/server/src/init.ts
Comment (akosyakov) [inline components/server/src/init.ts:113]: I'm not sure about these changes, since there can be important operations which should complete without leaving system in undefined state, i.e. requests in run to start workspace, rolling back some spice db changes, cascade deleting operations which don't have overall transaction, and so on. Stopping accepting new connections and letting event loop to run to the end seems to be desirable. 

Otherwise I think we can land it.

**Review by akosyakov — APPROVED** [review]
I've approved to unblock. I think the good part that we stop accepting new connectins.

/hold

because I have worries about process.exit given current code

File: components/server/src/init.ts
Comment (svenefftinge) [reply components/server/src/init.ts:113]: Those are all problems that exist, but it doesn'T really matter if we kill everyone without notice 30 seconds later.

Comment (svenefftinge) [reply components/server/src/init.ts:113]: It would be good, though if we can be sure that all participants on the stop signal await

Comment (akosyakov) [reply components/server/src/init.ts:113]: true given that in prod we still accept new connections

**Pull request merged**
